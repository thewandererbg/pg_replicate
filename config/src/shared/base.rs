use thiserror::Error;

/// Errors that can occur during configuration validation.
#[derive(Debug, Error)]
pub enum ValidationError {
    /// TLS is enabled but no trusted root certificates are provided.
    #[error("Invalid TLS config: `trusted_root_certs` must be set when `enabled` is true")]
    MissingTrustedRootCerts,

    /// General configuration validation error.
    #[error("Invalid configuration: {0}")]
    InvalidConfig(String),

    /// No destinations configured.
    #[error("At least one destination must be configured")]
    NoDestinations,

    /// Duplicate destination names or conflicting configurations.
    #[error("Duplicate or conflicting destination configuration: {0}")]
    DuplicateDestination(String),

    /// Invalid destination configuration.
    #[error("Invalid destination configuration: {0}")]
    InvalidDestination(String),
}
