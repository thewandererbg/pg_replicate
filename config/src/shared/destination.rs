use serde::{Deserialize, Serialize};
use std::fmt;

use crate::shared::ValidationError;

/// Configuration options for supported data destinations.
///
/// This enum is used to specify the destination type and its configuration
/// for the replicator. Variants correspond to different supported destinations.
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DestinationConfig {
    /// In-memory destination for ephemeral or test data.
    Memory,
    /// Google BigQuery destination configuration.
    ///
    /// Use this variant to configure a BigQuery destination, including
    /// project and dataset identifiers, service account credentials, and
    /// optional staleness settings.
    BigQuery {
        /// Google Cloud project identifier.
        project_id: String,
        /// BigQuery dataset identifier.
        dataset_id: String,
        /// Service account key for authenticating with BigQuery.
        gcp_sa_key_path: String,
        /// Maximum staleness in minutes for BigQuery CDC reads.
        max_staleness_mins: u16,
        /// Batch processing configuration.
        batch: BatchConfig,
    },
    /// Clickhouse destination configuration.
    ClickHouse {
        /// ClickHouse server URL
        url: String,
        /// ClickHouse database name
        database: String,
        /// ClickHouse username
        username: String,
        /// ClickHouse password
        password: String,
        /// Batch processing configuration.
        batch: BatchConfig,
    },
}

impl fmt::Debug for DestinationConfig {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Memory => f.write_str("Memory"),
            Self::BigQuery {
                project_id,
                dataset_id,
                gcp_sa_key_path: _,
                max_staleness_mins,
                batch,
            } => f
                .debug_struct("BigQuery")
                .field("project_id", project_id)
                .field("dataset_id", dataset_id)
                .field("gcp_sa_key_path", &"REDACTED")
                .field("max_staleness_mins", max_staleness_mins)
                .field("batch", batch)
                .finish(),
            Self::ClickHouse {
                url,
                database,
                username,
                password: _,
                batch,
            } => f
                .debug_struct("ClickHouse")
                .field("url", url)
                .field("database", database)
                .field("username", username)
                .field("password", &"REDACTED")
                .field("batch", batch)
                .finish(),
        }
    }
}

impl Default for DestinationConfig {
    fn default() -> Self {
        Self::Memory
    }
}

/// Batch processing configuration for pipelines.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct BatchConfig {
    /// Maximum number of items in a batch for table copy and event streaming.
    pub max_size: usize,
    /// Maximum time, in milliseconds, to wait for a batch to fill before processing.
    pub max_fill_ms: u64,
}

impl Default for BatchConfig {
    fn default() -> Self {
        Self {
            max_size: 1000,
            max_fill_ms: 1000,
        }
    }
}

impl BatchConfig {
    /// Validates the batch configuration
    pub fn validate(&self) -> Result<(), ValidationError> {
        if self.max_size == 0 {
            return Err(ValidationError::InvalidDestination(
                "Batch max_size must be greater than 0".to_string(),
            ));
        }
        if self.max_fill_ms == 0 {
            return Err(ValidationError::InvalidDestination(
                "Batch max_fill_ms must be greater than 0".to_string(),
            ));
        }
        Ok(())
    }
}
