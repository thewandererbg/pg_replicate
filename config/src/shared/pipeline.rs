use serde::{Deserialize, Serialize};

/// Configuration for a pipeline's batching and worker retry behavior.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct PipelineConfig {
    /// The unique identifier for this pipeline.
    ///
    /// A pipeline id determines isolation between pipelines, in terms of replication slots and state
    /// store.
    pub id: u64,
    /// Name of the Postgres publication to use for logical replication.
    pub publication_name: String,
}
