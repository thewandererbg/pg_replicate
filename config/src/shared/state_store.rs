use serde::{Deserialize, Serialize};

/// Configurations options for the state store.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StateStoreConfig {
    /// The in-memory state store.
    Memory,
}

impl Default for StateStoreConfig {
    fn default() -> Self {
        Self::Memory
    }
}
