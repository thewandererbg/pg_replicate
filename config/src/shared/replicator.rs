use crate::shared::pipeline::PipelineConfig;
use crate::shared::{
    BatchConfig, DestinationConfig, SourceConfig, StateStoreConfig, ValidationError,
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;

/// Configuration for the replicator service.
///
/// This struct aggregates all configuration required to run the replicator, including source,
/// destinations, pipeline, state store, and optional Supabase-specific settings.
///
/// The [`ReplicatorConfig`] is typically deserialized from a configuration file and passed to the
/// replicator at startup.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ReplicatorConfig {
    /// Configuration for the source Postgres instance.
    pub source: SourceConfig,
    /// Configuration for the state store used to persist replication state.
    pub state_store: StateStoreConfig,
    /// Configuration for replication destinations.
    pub destinations: DestinationsConfig,
    /// Configuration for the replication pipeline.
    pub pipeline: PipelineConfig,
}

/// Configuration for a specific destination type.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum DestinationTypeConfig {
    #[serde(rename = "clickhouse")]
    ClickHouse(ClickHouseConfig),
    #[serde(rename = "bigquery")]
    BigQuery(BigQueryConfig),
    // Easy to add new destination types here in the future
    // #[serde(rename = "kafka")]
    // Kafka(KafkaConfig),
}

/// Configuration for multiple named destinations with environment variable support.
///
/// This uses a type alias for HashMap to enable proper serialization/deserialization
/// while maintaining compatibility with environment variable overrides.
pub type DestinationsConfig = HashMap<String, DestinationTypeConfig>;

/// ClickHouse destination configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClickHouseConfig {
    /// ClickHouse server URL
    pub url: String,
    /// ClickHouse database name
    pub database: String,
    /// ClickHouse username
    pub username: String,
    /// ClickHouse password
    pub password: String,
    /// Batch processing configuration.
    #[serde(default)]
    pub batch: BatchConfig,
}

impl ClickHouseConfig {
    /// Validates the ClickHouse configuration
    pub fn validate(&self) -> Result<(), ValidationError> {
        if self.url.is_empty() {
            return Err(ValidationError::InvalidDestination(
                "ClickHouse URL cannot be empty".to_string(),
            ));
        }

        // Basic URL format validation
        if !self.url.starts_with("http://") && !self.url.starts_with("https://") {
            return Err(ValidationError::InvalidDestination(
                "ClickHouse URL must start with http:// or https://".to_string(),
            ));
        }

        if self.database.is_empty() {
            return Err(ValidationError::InvalidDestination(
                "ClickHouse database cannot be empty".to_string(),
            ));
        }
        if self.username.is_empty() {
            return Err(ValidationError::InvalidDestination(
                "ClickHouse username cannot be empty".to_string(),
            ));
        }
        if self.password.is_empty() {
            return Err(ValidationError::InvalidDestination(
                "ClickHouse password cannot be empty".to_string(),
            ));
        }

        // Validate batch configuration
        self.batch.validate()?;

        Ok(())
    }
}

/// BigQuery destination configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BigQueryConfig {
    /// Google Cloud project identifier.
    pub project_id: String,
    /// BigQuery dataset identifier.
    pub dataset_id: String,
    /// Service account key for authenticating with BigQuery.
    pub gcp_sa_key_path: String,
    /// Maximum staleness in minutes for BigQuery CDC reads.
    pub max_staleness_mins: u16,
    /// Batch processing configuration.
    #[serde(default)]
    pub batch: BatchConfig,
}

impl BigQueryConfig {
    /// Validates the BigQuery configuration
    pub fn validate(&self) -> Result<(), ValidationError> {
        if self.project_id.is_empty() {
            return Err(ValidationError::InvalidDestination(
                "BigQuery project_id cannot be empty".to_string(),
            ));
        }
        if self.dataset_id.is_empty() {
            return Err(ValidationError::InvalidDestination(
                "BigQuery dataset_id cannot be empty".to_string(),
            ));
        }
        if self.gcp_sa_key_path.is_empty() {
            return Err(ValidationError::InvalidDestination(
                "BigQuery gcp_sa_key_path cannot be empty".to_string(),
            ));
        }

        // Validate that the service account key file exists
        if !Path::new(&self.gcp_sa_key_path).exists() {
            return Err(ValidationError::InvalidDestination(format!(
                "BigQuery service account key file does not exist: {}",
                self.gcp_sa_key_path
            )));
        }

        // Validate reasonable range for staleness
        if self.max_staleness_mins > 1440 {
            // 24 hours
            return Err(ValidationError::InvalidDestination(
                "BigQuery max_staleness_mins should not exceed 1440 minutes (24 hours)".to_string(),
            ));
        }

        // Validate batch configuration
        self.batch.validate()?;

        Ok(())
    }
}

impl DestinationTypeConfig {
    /// Validates the destination configuration based on its type
    pub fn validate(&self) -> Result<(), ValidationError> {
        match self {
            DestinationTypeConfig::ClickHouse(config) => config.validate(),
            DestinationTypeConfig::BigQuery(config) => config.validate(),
        }
    }

    /// Converts the destination type config to a DestinationConfig enum
    pub fn to_destination_config(&self) -> DestinationConfig {
        match self {
            DestinationTypeConfig::ClickHouse(ch) => DestinationConfig::ClickHouse {
                url: ch.url.clone(),
                database: ch.database.clone(),
                username: ch.username.clone(),
                password: ch.password.clone(),
                batch: ch.batch.clone(),
            },
            DestinationTypeConfig::BigQuery(bq) => DestinationConfig::BigQuery {
                project_id: bq.project_id.clone(),
                dataset_id: bq.dataset_id.clone(),
                gcp_sa_key_path: bq.gcp_sa_key_path.clone(),
                max_staleness_mins: bq.max_staleness_mins,
                batch: bq.batch.clone(),
            },
        }
    }
}

/// Extension trait to add helper methods to DestinationsConfig (HashMap).
///
/// This trait provides all the helper methods that were previously on the struct,
/// allowing for clean method chaining and better ergonomics.
pub trait DestinationsConfigExt {
    /// Get all configured destinations as a vector of DestinationConfig enums.
    /// This is the primary method for accessing destinations.
    fn to_destination_configs(&self) -> Vec<DestinationConfig>;

    /// Get destination configurations with their names as a vector of tuples.
    fn get_named_destinations(&self) -> Vec<(String, DestinationConfig)>;

    /// Get all destination names as a vector.
    fn get_destination_names(&self) -> Vec<&String>;

    /// Check if any ClickHouse destinations are configured.
    fn has_clickhouse(&self) -> bool;

    /// Check if any BigQuery destinations are configured.
    fn has_bigquery(&self) -> bool;

    /// Get all ClickHouse destinations with their names.
    fn get_clickhouse_destinations(&self) -> Vec<(&String, &ClickHouseConfig)>;

    /// Get all BigQuery destinations with their names.
    fn get_bigquery_destinations(&self) -> Vec<(&String, &BigQueryConfig)>;

    /// Validates all configured destinations.
    fn validate(&self) -> Result<(), ValidationError>;
}

impl DestinationsConfigExt for DestinationsConfig {
    fn to_destination_configs(&self) -> Vec<DestinationConfig> {
        self.values()
            .map(|dest| dest.to_destination_config())
            .collect()
    }

    fn get_named_destinations(&self) -> Vec<(String, DestinationConfig)> {
        self.iter()
            .map(|(name, dest)| (name.clone(), dest.to_destination_config()))
            .collect()
    }

    fn get_destination_names(&self) -> Vec<&String> {
        self.keys().collect()
    }

    fn has_clickhouse(&self) -> bool {
        self.values()
            .any(|dest| matches!(dest, DestinationTypeConfig::ClickHouse(_)))
    }

    fn has_bigquery(&self) -> bool {
        self.values()
            .any(|dest| matches!(dest, DestinationTypeConfig::BigQuery(_)))
    }

    fn get_clickhouse_destinations(&self) -> Vec<(&String, &ClickHouseConfig)> {
        self.iter()
            .filter_map(|(name, dest)| match dest {
                DestinationTypeConfig::ClickHouse(config) => Some((name, config)),
                _ => None,
            })
            .collect()
    }

    fn get_bigquery_destinations(&self) -> Vec<(&String, &BigQueryConfig)> {
        self.iter()
            .filter_map(|(name, dest)| match dest {
                DestinationTypeConfig::BigQuery(config) => Some((name, config)),
                _ => None,
            })
            .collect()
    }

    fn validate(&self) -> Result<(), ValidationError> {
        if self.is_empty() {
            return Err(ValidationError::NoDestinations);
        }

        // Validate each destination
        for (name, dest) in self {
            dest.validate().map_err(|_| {
                // Don't double-wrap the error, just add context about which destination failed
                ValidationError::InvalidDestination(format!(
                    "Destination '{}' failed validation",
                    name
                ))
            })?;
        }

        Ok(())
    }
}

impl ReplicatorConfig {
    /// Get all configured destinations.
    /// This is the primary method for accessing destinations from ReplicatorConfig.
    pub fn get_destinations(&self) -> Vec<DestinationConfig> {
        self.destinations.to_destination_configs()
    }

    /// Get destination configurations with their names.
    pub fn get_named_destinations(&self) -> Vec<(String, DestinationConfig)> {
        self.destinations.get_named_destinations()
    }

    /// Get the number of configured destinations.
    pub fn destination_count(&self) -> usize {
        self.destinations.len()
    }

    /// Check if multiple destinations are configured.
    pub fn has_multiple_destinations(&self) -> bool {
        self.destinations.len() > 1
    }

    /// Check if there are no destinations configured.
    pub fn is_empty(&self) -> bool {
        self.destinations.is_empty()
    }

    /// Get a specific destination by name.
    pub fn get_destination(&self, name: &str) -> Option<&DestinationTypeConfig> {
        self.destinations.get(name)
    }

    /// Validates the loaded [`ReplicatorConfig`].
    ///
    /// Checks the validity of the TLS configuration and validates all destination configurations.
    ///
    /// # Errors
    ///
    /// Returns [`ValidationError`] if validation fails.
    pub fn validate(&self) -> Result<(), ValidationError> {
        // Validate TLS configuration
        self.source.tls.validate()?;

        // Validate destinations configuration
        self.destinations.validate()?;

        Ok(())
    }
}
