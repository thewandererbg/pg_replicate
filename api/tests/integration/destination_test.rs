use config::shared::DestinationConfig;
use reqwest::StatusCode;

use crate::{
    common::test_app::{
        spawn_test_app, CreateDestinationRequest, CreateDestinationResponse, DestinationResponse,
        DestinationsResponse, TestApp, UpdateDestinationRequest,
    },
    integration::tenants_test::create_tenant,
};

pub fn new_name() -> String {
    "BigQuery Destination".to_string()
}

pub fn new_destination_config() -> DestinationConfig {
    DestinationConfig::BigQuery {
        project_id: "project-id".to_string(),
        dataset_id: "dataset-id".to_string(),
        service_account_key: "service-account-key".to_string(),
        max_staleness_mins: None,
    }
}

pub fn updated_name() -> String {
    "BigQuery Destination (Updated)".to_string()
}

pub fn updated_destination_config() -> DestinationConfig {
    DestinationConfig::BigQuery {
        project_id: "project-id-updated".to_string(),
        dataset_id: "dataset-id-updated".to_string(),
        service_account_key: "service-account-key-updated".to_string(),
        max_staleness_mins: Some(10),
    }
}

pub async fn create_destination_with_config(
    app: &TestApp,
    tenant_id: &str,
    name: String,
    config: DestinationConfig,
) -> i64 {
    let destination = CreateDestinationRequest { name, config };
    let response = app.create_destination(tenant_id, &destination).await;
    let response: CreateDestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    response.id
}

pub async fn create_destination(app: &TestApp, tenant_id: &str) -> i64 {
    create_destination_with_config(app, tenant_id, new_name(), new_destination_config()).await
}

#[tokio::test(flavor = "multi_thread")]
async fn destination_can_be_created() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let destination = CreateDestinationRequest {
        name: new_name(),
        config: new_destination_config(),
    };
    let response = app.create_destination(tenant_id, &destination).await;

    // Assert
    assert!(response.status().is_success());
    let response: CreateDestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, 1);
}

#[tokio::test(flavor = "multi_thread")]
async fn an_existing_destination_can_be_read() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    let destination = CreateDestinationRequest {
        name: new_name(),
        config: new_destination_config(),
    };
    let response = app.create_destination(tenant_id, &destination).await;
    let response: CreateDestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let destination_id = response.id;

    // Act
    let response = app.read_destination(tenant_id, destination_id).await;

    // Assert
    assert!(response.status().is_success());
    let response: DestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, destination_id);
    assert_eq!(&response.tenant_id, tenant_id);
    assert_eq!(response.name, destination.name);
    insta::assert_debug_snapshot!(response.config);
}

#[tokio::test(flavor = "multi_thread")]
async fn a_non_existing_destination_cant_be_read() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let response = app.read_destination(tenant_id, 42).await;

    // Assert
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn an_existing_destination_can_be_updated() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    let destination = CreateDestinationRequest {
        name: new_name(),
        config: new_destination_config(),
    };
    let response = app.create_destination(tenant_id, &destination).await;
    let response: CreateDestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let destination_id = response.id;

    // Act
    let updated_config = UpdateDestinationRequest {
        name: updated_name(),
        config: updated_destination_config(),
    };
    let response = app
        .update_destination(tenant_id, destination_id, &updated_config)
        .await;

    // Assert
    assert!(response.status().is_success());
    let response = app.read_destination(tenant_id, destination_id).await;
    let response: DestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, destination_id);
    assert_eq!(&response.tenant_id, tenant_id);
    assert_eq!(response.name, updated_config.name);
    insta::assert_debug_snapshot!(response.config);
}

#[tokio::test(flavor = "multi_thread")]
async fn a_non_existing_destination_cant_be_updated() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let updated_config = UpdateDestinationRequest {
        name: updated_name(),
        config: updated_destination_config(),
    };
    let response = app.update_destination(tenant_id, 42, &updated_config).await;

    // Assert
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn an_existing_destination_can_be_deleted() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    let destination = CreateDestinationRequest {
        name: new_name(),
        config: new_destination_config(),
    };
    let response = app.create_destination(tenant_id, &destination).await;
    let response: CreateDestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let destination_id = response.id;

    // Act
    let response = app.delete_destination(tenant_id, destination_id).await;

    // Assert
    assert!(response.status().is_success());
    let response = app.read_destination(tenant_id, destination_id).await;
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn a_non_existing_destination_cant_be_deleted() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let response = app.delete_destination(tenant_id, 42).await;

    // Assert
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn all_destinations_can_be_read() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;
    let destination1_id =
        create_destination_with_config(&app, tenant_id, new_name(), new_destination_config()).await;
    let destination2_id = create_destination_with_config(
        &app,
        tenant_id,
        updated_name(),
        updated_destination_config(),
    )
    .await;

    // Act
    let response = app.read_all_destinations(tenant_id).await;

    // Assert
    assert!(response.status().is_success());
    let response: DestinationsResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    for destination in response.destinations {
        if destination.id == destination1_id {
            let name = new_name();
            assert_eq!(&destination.tenant_id, tenant_id);
            assert_eq!(destination.name, name);
            insta::assert_debug_snapshot!(destination.config);
        } else if destination.id == destination2_id {
            let name = updated_name();
            assert_eq!(&destination.tenant_id, tenant_id);
            assert_eq!(destination.name, name);
            insta::assert_debug_snapshot!(destination.config);
        }
    }
}
