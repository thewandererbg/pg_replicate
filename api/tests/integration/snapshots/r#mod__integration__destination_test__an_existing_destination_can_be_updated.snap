---
source: api/tests/integration/destination_test.rs
expression: updated_config.config
---
BigQuery {
    project_id: "project-id-updated",
    dataset_id: "dataset-id-updated",
    service_account_key: "REDACTED",
    max_staleness_mins: Some(
        10,
    ),
}
